gk/co.h

```diff
@@ -0,0 +1,184 @@
+/*
+ * Gatekeeper - DoS protection system.
+ * Copyright (C) 2016 Digirati LTDA.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef _GATEKEEPER_GK_CO_H_
+#define _GATEKEEPER_GK_CO_H_
+
+#include <stdbool.h>
+#include <string.h>
+#include <coro.h>
+#include <list.h>
+
+#include "gatekeeper_gk.h"
+#include "gatekeeper_acl.h"
+
+/* Information transfered between tasks assigned to the same coroutine. */
+struct gk_co_leftover {
+	struct flow_entry *fe;
+	struct gk_fib     *fib;
+	/* @temp_fe is only in use when @fe == &temp_fe. */
+	struct flow_entry temp_fe;
+};
+
+struct gk_co {
+	/*
+	 * Attach this coroutine to work->working_cos while
+	 * this coroutine is working. Otherwise, attach to work->idle_cos.
+	 */
+	struct list_head    cos_list;
+	/* structs from libcoro. */
+	struct coro_stack   stack;
+	struct coro_context coro;
+	/* Task assigned to this coroutine. */
+	struct list_head    task_queue;
+	struct gk_co_work   *work;
+};
+
+struct gk_co_task *task;
+
+typedef void (*gk_co_task_func_t)(struct gk_co *this_co,
+	struct gk_co_task *task, struct gk_co_leftover *leftover);
+
+struct gk_co_task {
+	/*
+	 * Initially attach this task to work->tasks and,
+	 * once the task is assigned to a coroutine, attach this task to
+	 * co->task_queue.
+	 */
+	struct list_head  task_list;
+	uint32_t          task_hash;
+	void              *task_arg;
+	gk_co_task_func_t task_func;
+};
+
+struct gk_co_work {
+	/* All tasks to be done. */
+	struct list_head tasks;
+	/* The coroutines working on the tasks. */
+	struct list_head working_cos;
+	/*
+	 * Idle coroutines; there is no task for them.
+	 * They become working coroutines when new tasks are added or
+	 * rescheduled.
+	 */
+	struct list_head idle_cos;
+	/*
+	 * Map from a masked task hash to the assigned coroutine.
+	 *
+	 * This is important to avoid race conditions between coroutines.
+	 * For example, assume that two tasks that are going to work on
+	 * the same flow entry are assigned to two different coroutines, and
+	 * that the corresponding flow entry is not available in
+	 * the flow table, both coroutines may try to add the same flow entry.
+	 * If these two tasks share the same hash, this map guarantees that
+	 * both tasks are going to be assigned to the same coroutine.
+	 */
+	struct gk_co     **hash_to_assigned_co;
```

Comment:
This field allows us to map unassigned tasks to coroutines dynamically, which can better load balance the tasks among coroutines while ensuring tasks from the same flow are processed by only one coroutine at any time. This way it guarantees the in-order packet delivery in a flow. However, Gatekeeper may have to go over the tasks list to find the first unassigned task by calling ``get_the_very_first_task()``, which could become the bottleneck even the cost for transferring between coroutines is very small.

An alternative solution is to simply assign tasks to coroutine by using the ``task_hash`` field in ``struct gk_co_task``, which actually is similar to the RSS. If so, we can maintain a list of assigned tasks and a list of unassigned tasks. Each time, ``get_the_very_first_task()`` simply return a tasks from the later list. If there is no unassigned tasks, it simply transfer to the root coroutine.

Reply:
`struct gk_co` already has the field `struct list_head task_queue` for assigned tasks. And `work->tasks` holds unsigned tasks. After initialization, `gk_co_main()` calls `next_task()` which gets tasks from the field `struct list_head task_queue` and falls back to `work->tasks` if needed. So the code already does what you are proposing.

The reason the field `struct list_head task_queue` is often empty in our evaluation is that our evaluation rarely has packets of the same flow, especially within a batch.

Now, if you meant assigning all tasks before putting the coroutines to work. This might be more efficient and we can test it. But I don't see from where this efficiency would come.

Reply:
The new code assigns the tasks to coroutines before the coroutines begin to run as you suggested. Could you evaluate the code? Please add counter, CPU, and load statistics for 3 lcores and a 2^25-entry flow table.

Reply:
Thanks @AltraMayor ! Sure, will do it today.